src/core/datatypes/trees/TopologyNode.cpp

```diff
@@ -2012,48 +1978,95 @@ void TopologyNode::setUseAges(bool tf, bool recursive)
+        //if ( isRoot() == false )
+        //{
 
-}
+            if ( getNumberOfChildren() == 1 )
+            {
 
+                /* Should we remove the node or replace it by a bifurcation with a zero-length branch instead?
+                 * I.e.,
+                 *
+                 *    A                   A                 A
+                 *    |                   |                 |
+                 *    |                   |                 |
+                 *    B      -->      B --+--\      or      |
+                 *    |                      |              |
+                 *    |                      |              |
+                 *    C                      C              C
+                 *
+                 *                 (replace = true)  (replace = false)
+                 */
+                if ( replace == true )
+                {
+                    TopologyNode *new_fossil = new TopologyNode( getTaxon() );
+                    taxon = Taxon("");
 
-void TopologyNode::setTaxon(Taxon const &t)
-{
+                    // connect to the old fossil
+                    addChild( new_fossil );
+                    new_fossil->setParent( this );
 
-    taxon = t;
+                    // set the fossil flags
+                    setSampledAncestor( false );
+                    new_fossil->setSampledAncestor( true );
 
-}
+                    // set the age and branch-length of the fossil
+                    new_fossil->setAge( age );
+                    new_fossil->setBranchLength( 0.0 );
+                }
+                else
+                {
+                    // we are going to delete myself by connect my parent and my child
+                    TopologyNode& parent = getParent();
+                    TopologyNode& child = getChild(0);
+                    
+                    // the new branch length needs to be the branch length of the parent and child
+                    double summ = getBranchLength() + child.getBranchLength();
+                    
+                    // now remove myself from the parent
+                    parent.removeChild( this );
+                    
+                    // and my child from me
+                    removeChild( &child );
+                    
+                    // and stich my parent and my child together
+                    parent.addChild( &child );
+                    child.setParent( &parent );
+                    
+                    // finally, adapt the branch lengths
+                    child.setBranchLength(summ);
+                    
+                }
 
 
-void TopologyNode::setTimeInStates(std::vector<double> t)
-{
-    time_in_states = t;
-}
+            }
 
+        //}
 
-void TopologyNode::setTree(Tree *t)
-{
+        // call this function recursively for all its children
+        for (size_t i=0; i<getNumberOfChildren(); ++i)
+        {
+            getChild( i ).suppressOutdegreeOneNodes( replace );
```

Comment:
does that work if replace = false ? since we've just removed ourselves from the tree, can we still recurse ?

Reply:
Good catch! Indeed we cannot; we get a segfault if we try. I've fixed this now – the recursion only takes place when `replace = true`. When it's false, the removal of outdegree-1 nodes is instead handled by a combination of iteration and recursion at the level of `Tree::suppressOutdegreeOneNodes`. I've added an integration test to make sure that we can handle trees with multiple outdegree-1 nodes (including potentially at the root) under both `replace = true` and `replace = false`.